metrics/src/main/java/com/expedia/adaptivealerting/metrics/functions/MetricFunctionsApp.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2018-2019 Expedia Group, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expedia.adaptivealerting.metrics.functions;
+
+import com.expedia.adaptivealerting.kafka.TypesafeConfigLoader;
+import com.expedia.adaptivealerting.kafka.util.ConfigUtil;
+import com.expedia.adaptivealerting.metrics.functions.source.MetricFunctionsReader;
+import com.expedia.adaptivealerting.metrics.functions.source.MetricFunctionsSpec;
+import com.expedia.metrics.MetricData;
+import lombok.val;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class MetricFunctionsApp {
+
+    private static final String METRIC_SOURCE_SINK = "metric-source-sink";
+    private static final String APP_ID = "aa-metric-functions";
+    private static final String AGGREGATOR_PRODUCER = "aggregator-producer";
+    private static final String INPUT_FUNCTIONS_FILENAME ="functions.txt";
+    private final static String INPUT_FILE_PATH = "/config/";
+    private static final int corePoolSize = 5;
+
+    public static void main(String[] args) {
```

Comment:
The input to the module is not kafka, so this cannot be done.

Reply:
@dinilatgit -- I don't follow your response to @karanshah03 's suggestion.

The idea is that there's a core logic to metric functions that has nothing to do with Kafka. So the input to your module would be independent of the messaging implementation. So couldn't you take the Kafka-specific part and move that to the Kafka module?

Another way to think of this is what happens when we want to add support for Kinesis? Do we have to reimplement your entire metrics xform code for Kinesis? It would be better if they could just implement a Kinesis wrapper for the core logic.

Reply:
created KafkaMetricFunctions for the decoupling request.

Reply:
Let's pair on this one.

Reply:
OK looks like we won't have time to sync today, so let me describe what I have in mind. There should be two general pieces here -- a core metric functions component and a Kafka wrapper. The Kafka wrapper should call the metric functions and not vice versa -- we want to be able to add other wrappers like Kinesis and so that means that the core metric functions can't know about its messaging environment impl.

This leaves open the question of how the metric functions comp is supposed to publish its metrics to Kafka if it doesn't know it's running in Kafka. To solve this, create a generic callback interface for publishing metrics. You can call it `TransformedMetricTopic` or similar. It should have a single method that allows you to publish the metric in question, and the method shouldn't know about Kafka.

Then when you initialize the Kafka wrapper, implement the callback interface and pass that into the core metric functions component. That gives the component a way to publish metrics to Kafka without having to know anything about Kafka.